torchao/csrc/cpu/int8_sdpa.cpp

```diff
@@ -0,0 +1,2382 @@
+  }
+}
+
+template <typename scalar_t>
+inline void _int_sum_a_contiguous_kernel(
+    const scalar_t* in,
+    int32_t* out,
+    const int& M,
+    const int& N,
+    const int& ld,
+    const int32_t& scale) {
+  const int32_t vec_size = at::vec::Vectorized<int32_t>::size();
+  auto vec_scale = at::vec::Vectorized<int32_t>(scale);
+  // initialization with 0
+  int32_t zero = 0;
+  auto vec_zero = at::vec::Vectorized<int32_t>(zero);
+  for (long i = 0; i < vec_size * (M / vec_size); i += vec_size) {
+    _store(out + i, vec_zero);
+  }
+  for (long i = vec_size * (M / vec_size); i < M; i++) {
+    out[i] = zero;
+  }
+  // sum
+  for (long j = 0; j < N; j++) {
+    const scalar_t* tmp_in = in + j * ld;
+    for (long i = 0; i < vec_size * (M / vec_size); i += vec_size) {
+      auto tmp0 = at::vec::Vectorized<scalar_t>::loadu(tmp_in + i);
+      auto tmp1 = at::vec::Vectorized<int32_t>::loadu(out + i);
+      auto tmp2 = at::vec::convert<int32_t>(tmp0);
+      auto tmp3 = tmp1 + tmp2;
+      _store(out + i, tmp3);
+    }
+    for (long i = vec_size * (M / vec_size); i < M; i++) {
+      auto tmp0 = tmp_in[i];
+      auto tmp1 = out[i];
+      auto tmp2 = static_cast<int32_t>(tmp0);
+      auto tmp3 = tmp1 + tmp2;
+      out[i] = tmp3;
+    }
+  }
+  // scale
+  for (long i = 0; i < vec_size * (M / vec_size); i += vec_size) {
+    auto tmp0 = at::vec::Vectorized<int32_t>::loadu(out + i);
+    auto tmp1 = tmp0 * vec_scale;
+    _store(out + i, tmp1);
+  }
+  for (long i = vec_size * (M / vec_size); i < M; i++) {
+    auto tmp0 = out[i];
+    auto tmp1 = tmp0 * scale;
+    out[i] = tmp1;
+  }
+}
+
+inline void do_convert_u8_s8(
+    unsigned char* src,
+    signed char* dst,
+    int64_t in_rows,
+    int64_t in_cols,
+    int64_t ldi,
+    int64_t ldo) {
+  auto vec_size = at::vec::Vectorized<int16_t>::size();
+  auto vec_128 = at::vec::Vectorized<int16_t>(128);
+  for (int64_t r = 0; r < in_rows; r++) {
+    const unsigned char* tmp_src = src + r * ldi;
+    signed char* tmp_dst = dst + r * ldo;
+    for (int64_t c = 0; c < vec_size * (in_cols / vec_size); c += vec_size) {
+      auto tmp0 = at::vec::Vectorized<unsigned char>::loadu(tmp_src + c, vec_size);
+      auto tmp1 = at::vec::convert<int16_t>(tmp0);
+      auto tmp2 = tmp1 - vec_128;
+      auto tmp3 = at::vec::convert<signed char>(tmp2);
+      _store(tmp_dst + c, tmp3, vec_size);
+    }
+    for (int64_t c = vec_size * (in_cols / vec_size); c < in_cols; c++) {
+      auto tmp0 = tmp_src[c];
+      auto tmp1 = (int16_t) tmp0;
+      auto tmp2 = tmp1 - 128;
+      auto tmp3 = (signed char) tmp2;
+      tmp_dst[c] = tmp3;
+    }
+  }
+}
+
+template <typename scalar_t>
+inline void do_transpose(
+    scalar_t* src,
+    scalar_t* dst,
+    int64_t in_rows,
+    int64_t in_cols,
+    int64_t ldi,
+    int64_t ldo) {
+  for (int64_t r=0; r<in_rows; r++) {
+    for (int64_t c=0; c<in_cols; c++) {
+      *(dst + c * ldo + r) = *(src + r * ldi + c);
+    }
+  }
+}
```

Comment:
nit: Does compiler auto-vectorization work well in this case? I'm guessing you've already determined that this code isn't significant enough to optimize. Just curious.

Thanks

Reply:
Thanks! The transposition code is not significantly time-consuming, so this is currently a scalar version. The vectorized code would be added in the future if necessary.

Reply:
OK, thanks!